test/mk/auto.mk

```diff
@@ -1,29 +1,113 @@
-else ifeq ($(HOST_PLATFORM),Linux-aarch64)
-	CFLAGS += -DMLK_FORCE_AARCH64
+# Linux: Use /proc/cpuinfo
+MK_HOST_SUPPORTS_AVX2 := $(call check_host_feature,avx2,cat /proc/cpuinfo)
+MK_HOST_SUPPORTS_SSE2 := $(call check_host_feature,sse2,cat /proc/cpuinfo)
+MK_HOST_SUPPORTS_BMI2 := $(call check_host_feature,bmi2,cat /proc/cpuinfo)
+else ifeq ($(HOST_PLATFORM),Darwin-x86_64)
+# macOS: Use sysctl
+MK_HOST_SUPPORTS_AVX2 := $(call check_host_feature,AVX2,sysctl -n machdep.cpu.leaf7_features)
+MK_HOST_SUPPORTS_SSE2 := $(call check_host_feature,SSE2,sysctl -n machdep.cpu.features)
+MK_HOST_SUPPORTS_BMI2 := $(call check_host_feature,BMI2,sysctl -n machdep.cpu.leaf7_features)
+else ifneq ($(CROSS_PREFIX),)
+# Cross-compilation: assume all features are supported
+MK_HOST_SUPPORTS_AVX2 := 1
+MK_HOST_SUPPORTS_SSE2 := 1
+MK_HOST_SUPPORTS_BMI2 := 1
+else
+# Other platforms: assume no support
+MK_HOST_SUPPORTS_AVX2 := 0
+MK_HOST_SUPPORTS_SSE2 := 0
+MK_HOST_SUPPORTS_BMI2 := 0
+endif # HOST_PLATFORM x86_64
+
+endif # x86_64
+
+# AArch64 architecture detection
+ifeq ($(ARCH),aarch64)
+
+# Host CPU feature detection for AArch64
+ifeq ($(HOST_PLATFORM),Linux-aarch64)
+# Linux: Use /proc/cpuinfo (look for sha3 in Features line)
+MK_HOST_SUPPORTS_SHA3 := $(call check_host_feature,sha3,cat /proc/cpuinfo)
 else ifeq ($(HOST_PLATFORM),Darwin-arm64)
-	CFLAGS += -DMLK_FORCE_AARCH64
+# macOS: Use sysctl to check for SHA3 support
+MK_HOST_SUPPORTS_SHA3 := $(call check_host_feature,1,sysctl -n hw.optional.armv8_2_sha3)
+else ifneq ($(CROSS_PREFIX),)
+# Cross-compilation: assume all features are supported
+MK_HOST_SUPPORTS_SHA3 := 1
+else
+# Other platforms: assume no support
+MK_HOST_SUPPORTS_SHA3 := 0
+endif # HOST_PLATFORM aarch64
+
+endif # aarch64
+
+# Only apply CFLAGS modifications if AUTO=1
+ifeq ($(AUTO),1)
+
+# x86_64 CFLAGS configuration
+ifeq ($(ARCH),x86_64)
+CFLAGS += -DMLK_FORCE_X86_64
+
+# Add flags only if both compiler and host support the feature
+ifeq ($(MK_COMPILER_SUPPORTS_AVX2)$(MK_HOST_SUPPORTS_AVX2),11)
+CFLAGS += -mavx2
 endif
-# Cross compilation
-else ifneq ($(findstring x86_64, $(CROSS_PREFIX)),)
-	CFLAGS += -mavx2 -mbmi2 -mpopcnt -maes
-	CFLAGS += -DMLK_FORCE_X86_64
-else ifneq ($(findstring aarch64_be, $(CROSS_PREFIX)),)
-	CFLAGS += -DMLK_FORCE_AARCH64_EB
-else ifneq ($(findstring aarch64, $(CROSS_PREFIX)),)
-	CFLAGS += -DMLK_FORCE_AARCH64
-else ifneq ($(findstring riscv64, $(CROSS_PREFIX)),)
-	CFLAGS += -DMLK_FORCE_RISCV64
-else ifneq ($(findstring riscv32, $(CROSS_PREFIX)),)
-	CFLAGS += -DMLK_FORCE_RISCV32
-else ifneq ($(findstring powerpc64le, $(CROSS_PREFIX)),)
-	CFLAGS += -DMLK_FORCE_PPC64LE
+
+ifeq ($(MK_COMPILER_SUPPORTS_BMI2)$(MK_HOST_SUPPORTS_BMI2),11)
+CFLAGS += -mbmi2
 endif
+endif # x86_64
+
+# AArch64 CFLAGS configuration
+ifeq ($(ARCH),aarch64)
+CFLAGS += -DMLK_FORCE_AARCH64
+
+# Add SHA3 flags only if both compiler and host support it
+ifeq ($(MK_COMPILER_SUPPORTS_SHA3)$(MK_HOST_SUPPORTS_SHA3),11)
+CFLAGS += -march=armv8.4-a+sha3
```

Comment:
I wonder if this overwrites user-provieded CFLAGS. For example, if I specifically call
```
 tests all --cflags="-march=armv8.4-a"
```
That ideally shouldn't be overwritten by the auto detection?

Reply:
Okay, there seems to be no good way around this. Maybe we at least consistently set `AUTO=0` in CI when we set -march= manually?

Reply:
Yes, when archflags are provided one shouldn't use auto. We could try to remove the `-march` flags from the CI -- it should not work correctly by itself... or so the theory.